drivers/CMakeLists.txt

```diff
@@ -43,3 +43,4 @@ add_subdirectory_ifdef(CONFIG_NET_L2_ETHERNET  ethernet)
 add_subdirectory_ifdef(CONFIG_ENTROPY_HAS_DRIVER entropy)
 add_subdirectory_ifdef(CONFIG_SYS_CLOCK_EXISTS timer)
 add_subdirectory_ifdef(CONFIG_NEURAL_NET_ACCEL neural_net)
+add_subdirectory_ifdef(CONFIG_PTP_CLOCK        ptp_clock)
```

Comment:
Reading description: "It is useful that the ptp_clock_get() function can be called from
the userspace." Instead seeing: whole new directory is added, in drivers/! What's so driver'ish in this stuff?

Reply:
> Instead seeing: whole new directory is added, in drivers/! What's so driver'ish in this stuff?

Earlier the functionality was in inline function in ptp_clock.h. Needs to place now the func into .c file but the file was missing so it is now created.

Reply:
> Needs to place now the func into .c file but the file was missing so it is now created.

Is it device-specific? If not, we have subsys/net/ for code independent of particular hardware.

Reply:
it is device specific, see ptp_clock API.

This new directory is just fine.

Reply:
> it is device specific, see ptp_clock API.

"It" - what? I'm talking about the code added to  drivers/ptp_clock/ in this PR. Tell me to which device it's specific and that it won't run on other devices.

Reply:
You don't seem to know much about device drivers: ptp_clock_get() is a device driver API, all about implementing such API*, as well as its userland exposure, should be in drivers/<device type>/

*: in the particular case of ptp, driver implementation are mostly (if not always, maybe there are ptp clock unrelated to ethernet) going to be made along with the relevant ethernet driver.

Reply:
> device driver API, all about implementing such API*

With such outlook, everything is "device driver", because any API in the end is implemented with device driver. Then you can put entire IP stack in drivers/ . 

The situation is (should be) very simple: a device driver is a piece of code which talks to one, and not another, piece of hardware. That should live in drivers/. Any device independent code, like userspace call shims for public APIs, as added in this patch, should not live in drivers/, because - surprise - they're not drivers.